addons/stock/models/stock_quant.py

```diff
@@ -281,6 +278,16 @@ def _update_reserved_quantity(self, product_id, location_id, quantity, lot_id=No
                 break
         return reserved_quants
 
+    @api.model
+    def _unlink_zero_quants(self):
+        """ _update_available_quantity do not unlink quant since it
+        invalidates the cache and reduces performance. This method
+        should be call by method that modify quant in order to perform
+        the unlink in batch and invalidate the cache only once.
+        """
```

Comment:
`_update_available_quantity` may leave quants with no quantity and no reserved_quantity. It
used to directly unlink these zero quants but this proved to hurt the performance as this
method is often called in batch and each unlink invalidate the cache. We defer the calls to
unlink in this method.